scripts/sort-makefile-am-sources.pl

```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/perl
+#
+## Copyright (C) 1996-2020 The Squid Software Foundation and contributors
+##
+## Squid software is distributed under GPLv2+ license and includes
+## contributions from numerous individuals and organizations.
+## Please see the COPYING and CONTRIBUTORS files for details.
+##
+
+use strict;
+use warnings;
+
+while (<>) {
+    if (m!^#!) {
+        print;
+        next;
+    }
+    if (/^(\S+_SOURCES)\s*=\s*\\$/) {
```

Comment:
To have fewer exceptions to worry about, I would cover all X_SOURCES assignments (and use a single line for single-item assignments).

I do not insist on this change if you have good reasons to exclude assignments that (currently) occupy one line or (in the future) start with a filename on the same line as X_SOURCES.

Reply:
Right now these will be skipped over. It would be possible to separate them into continuation-style assignments, but I'd leave that to a separate PR (if ever). The objective here is to ensure we're not double-including - which we are in master, btw

Reply:
> The objective here is to ensure we're not double-including

That does not sound like the right objective to me. The bad effects of duplicate entries are negligible AFAICT. I was under impression that the primary objective here is to reduce developer/reviewer time on finding, adding, and reordering list entries by ensurung One Canonical Order. That objective, technically, applies to all lists that have more than one entry.

In my experience, if a literal interpretation of a rule makes it applicable, the rule will be applied by some Squid reviewers. I recommend removing this potential source of future conflicts by ordering all SOURCE lists (which implies ordering all SOURCE lists that contain more than one item, including lists that are currently using a single line). Implementing the corresponding support should not be difficult AFAICT.

This change request falls into the `minor+moderate=?` [category](https://github.com/measurement-factory/squid-notes/blob/start/review.md#why-cleanup-now) in my decision table. To give you the benefit of that `?` doubt, I do not insist on this change if there are good reasons to delay it. So far, you have not provided such reasons AFAICT.

Reply:
Just to avoid misunderstanding since I am doing a fresh review: This old change request has not been addressed as of 55c6f83 and is incorporated into this new review.

Reply:
> > The objective here is to ensure we're not double-including
> 
> That does not sound like the right objective to me. The bad effects of duplicate entries are negligible AFAICT. I was under impression that the primary objective here is to reduce developer/reviewer time on finding, adding, and reordering list entries by ensurung One Canonical Order. That objective, technically, applies to all lists that have more than one entry.

Correct.

> In my experience, if a literal interpretation of a rule makes it applicable, the rule will be applied by some Squid reviewers. I recommend removing this potential source of future conflicts by ordering all SOURCE lists (which implies ordering all SOURCE lists that contain more than one item, including lists that are currently using a single line). Implementing the corresponding support should not be difficult AFAICT.

Doable, but the benefit is greatly reduced there

> This change request falls into the `minor+moderate=?` [category](https://github.com/measurement-factory/squid-notes/blob/start/review.md#why-cleanup-now) in my decision table. To give you the benefit of that `?` doubt, I do not insist on this change if there are good reasons to delay it. So far, you have not provided such reasons AFAICT.

I am happy with NODELAY :)

Reply:
Clarifying for context: current plan here is WONTFIX: single-line SOURCE files are simple enough that the benefits are not worth the extra complexity IMO

Reply:
Sorry I misunderstood your `NODELAY` comment. The clarified current plan does not address my change request.

Reply:
> IMO, you should revert _all_ manual changes to Makefiles

Yes, I will do this

Reply:
Okay, I've given in and implemented support for

X_SOURCES = foo.h foo.cc bar.h \
  bar.cc

Revision coming up as soon as I'm confident it will not break stuff.
It comes at a cost in code readability, unfortunately

Reply:
I may be able to help with readability.

Reply:
As of a3c06d9, the script does not seem to handle single-item SOURCE lists specially. AFAICT, such lists will now be split into two lines. This change request specifically asked to keep such lists on one line. Is that split an intentional formatting change? If it was an accidental formatting change, please fix.

Reply:
> They will be handled (To ne specific: passed through with no modification) by virtue of not matching the SOURCES regexp, because they don’t have the backslash at the end of the line. 

```Perl
     if (/^(\S+_SOURCES)\s*(\+?=)\s*(.*?)\s*\\$/) {
```

Aha, I think I found the source of the confusion: This change request was about _existing_ expressions like `...SOURCES = b.cc a.cc` that have multiple files, all on one line (and do not have the backslash at the end of the line):

```
ACL_REGISTRATION_SOURCES = AclRegs.cc AuthReg.cc
libfs_la_SOURCES = Module.cc Module.h
libheap_a_SOURCES = heap/store_heap_replacement.h heap/store_heap_replacement.cc heap/store_repl_heap.cc
debug_SOURCES = debug.cc stub_libmem.cc $(DEBUG_SOURCE) 
ESIExpressions_SOURCES = ESIExpressions.cc stub_libmem.cc $(DEBUG_SOURCE)
mem_node_test_SOURCES = mem_node_test.cc $(DEBUG_SOURCE)
mem_hdr_test_SOURCES = mem_hdr_test.cc $(DEBUG_SOURCE)
splay_SOURCES = splay.cc stub_libmem.cc $(DEBUG_SOURCE)
syntheticoperators_SOURCES = syntheticoperators.cc stub_libmem.cc $(DEBUG_SOURCE)
VirtualDeleteOperator_SOURCES = VirtualDeleteOperator.cc stub_libmem.cc $(DEBUG_SOURCE)
```

Commit 1ba2996 claimed to add support for "multiple files on single line", but that support was apparently limited to (potential/future?) expressions like `...SOURCES = b.cc a.cc \...` that do have a backslash at the end. I did not even realize that the latter were not covered before this change request. Thank you for fixing that!

Please remove that backslash termination requirement to cover the existing expressions like `...SOURCES = b.cc a.cc` that have multiple files, all on one line (and do not have the backslash at the end of the line) while rendering single-file lists on one line.

Reply:
> Please remove that backslash termination requirement to cover the existing expressions like `...SOURCES = b.cc a.cc` that have multiple files, all on one line (and do not have the backslash at the end of the line)
>  while rendering single-file lists on one line.

This last bit of rendering single-file lists on one line will render some lists quite ugly. Some have very long prefix names, some have automake variables as that single 'file' in the list. Most they occur in the unit test lists surrounded by other lines with multi-file indenting. So the one-line stands out oddly different.

IMO for *_SOURCES we should consistently have "SOURCES = \" then the file(s) on following lines indented with a single tab.

Reply:
FWIW, I am content with rendering single-file SOURCES lists on one line (my weak preference) or multiple lines (Amos' suggestion).

* The (small) advantages of Amos' solution are that it will allow for single-line diffs in more (future modification) cases and that it eliminates long single-line single-file assignments that look ugly.
* Its (small) disadvantages are that multi-line short-single-file assignment look odd/strange (to some readers) and that will increase this PR diff.

The best solution would probably enforce an 80-character line-length limit for single-file assignments (wrapping long lines and only them). However, I am worried that Francesco will run out of space for needles in my voodoo doll if I suggest that, so I will not ;-).

Reply:
Commit 6bdb11e187e5c97ba431b3b0190b63402befbd85 resolves this change request. I think the code can be simplified quite a bit, but I do not insist on that simplification. @kinkie, please let me know if you want me to simplify it.

@yadij, if you want multi-line assignments for single-line lists, please track that in a separate/new change request.

Reply:
> The best solution would probably enforce an 80-character line-length limit for single-file assignments (wrapping long lines and only them). However, I am worried that Francesco will run out of space for needles in my voodoo doll if I suggest that, so I will not ;-).

No need for needles - I'm missing the doll altogether. In general I believe we as a team are overthinking overdesigning things, and this decreases our collective ROTS (Return On Time Spent). Not something that can be addressed in a PR tho :P

I left single-line assignments in intentionally, as it is my preference. Not enough of a preference to prevent anyone from changing it, but enough of one that this is what I set in code.

Reply:
> I believe we as a team are overthinking overdesigning things, and this decreases our collective ROTS

Yes, I am aware of, but do not (unfortunately) share your belief. I believe that our _primary_ sins are underthinking and overresisting. There are certainly cases of overthinking and overdesigning as well (e.g., this comment of mine), but they are not the primary problem IMO. Naturally, it is very difficult to reconcile our two observations. I cannot suggest a solution beyond trying to agree on common [principles](https://github.com/measurement-factory/squid-notes/blob/start/README.md) and meeting more often.

On the bright side, we do agree on the outcome assessment -- low (and often negative) returns!